src/dmd/typesem.d

```diff
@@ -3624,6 +3624,23 @@ Expression dotExp(Type mt, Scope* sc, Expression e, Identifier ident, int flag)
             }
             e = e.castTo(sc, e.type.nextOf().pointerTo());
         }
+        else if (ident == Id._tupleof)
+        {
+            if (e.op == EXP.type)
+            {
+                e.error("Can only use .tupleof on static array values (perhaps you meant `typeof(%s.init.tupleof)`?)", mt.toChars);
+                return ErrorExp.get();
+            }
+            else
+            {
+                uinteger_t length = mt.dim.toUInteger();
+                auto exps = new Expressions();
+                exps.reserve(cast(size_t)length);
+                foreach (i; 0 .. length)
+                    exps.push(new IndexExp(e.loc, e, new IntegerExp(e.loc, i, Type.tsize_t)));
```

Comment:
Might be nice to avoid allocating IntegerExps on every iteration.

Reply:
I'm not sure how to do that, could you elaborate? (apologies if I'm asking the obvious here, honestly this is my first deep foray into dmd)

Reply:
If you know you have `length` IntegerExps to allocate you can allocate one buffer and then emplace into it.

Reply:
We don't do that anywhere else, I think it's ok as it is

Reply:
Yeah, I'm not sure what the right way would be to allocate the buffer, short of some ugly `__traits` and casting.